### PR TITLE
[FLINK-36883] Adapt time attributes from the VIEWs query

### PR DESCRIPTION
**Pull request by dawidwys**
## What is the purpose of the change

Fix Views so that they may have time attributes.

## Verifying this change

Added a unit test

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

**Comment by davidradl** [comment]
@dawidwys please could you correct the Jira id in the title - I was struggling to find it.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java [hunk 1]
```diff
@@ -149,12 +157,129 @@ void testResolvingProctimeOfCustomTableTableApi() throws Exception {
         testUtil.verifyExecPlan(table);
     }
 
+    @TestTemplate
+    void testTimeAttributeOfView() {
+        if (!streamingMode) {
+            // time attributes not supported in batch
+            return;
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java:165]: This is a test, please read the class again.

File: flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/catalog/SqlCatalogViewTable.java [hunk 2]
```diff
@@ -56,6 +60,62 @@ public SqlCatalogViewTable(
...
+                                + "]"
+                                + " castRowType ["
+                                + targetType
+                                + "]");
+            } else {
+                return adaptTimeAttributeInRecord(
+                        (RelRecordType) queryType, (RelRecordType) targetType, typeFactory);
+            }
+        } else {
+            return adaptTimeAttributeInSimpleType(queryType, targetType, typeFactory);
```
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/catalog/SqlCatalogViewTable.java:87]: 1. Time attributes can not be nested in other collection types
2. The logic is an adapted copy of Calcite's `RelDataTypeFactory#copyType`

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (davidradl) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java:165]: I realised this - so you think it is not work testing the time attributes for non streaming to show that is fails nicely?

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java [hunk 3]
```diff
@@ -149,12 +155,120 @@ void testResolvingProctimeOfCustomTableTableApi() throws Exception {
...
+                "CREATE TABLE t(i INT, ts TIMESTAMP_LTZ(3), WATERMARK FOR "
+                        + "ts AS ts) WITH ('connector' = 'datagen')");
+        tableEnvironment.executeSql("CREATE VIEW `cat`.`default`.v AS SELECT * FROM t");
+        testUtil.verifyExecPlan(
+                "SELECT sum(i), window_start "
+                        + "FROM TABLE(\n"
+                        + "   TUMBLE(\n"
+                        + "     DATA => TABLE `cat`.`default`.v,\n"
+                        + "     TIMECOL => DESCRIPTOR(ts),\n"
+                        + "     SIZE => INTERVAL '10' MINUTES))"
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java:177]: nit 
```suggestion
                        + "     SIZE => INTERVAL '10' MINUTES))\n"
```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java [hunk 4]
```diff
@@ -149,12 +155,120 @@ void testResolvingProctimeOfCustomTableTableApi() throws Exception {
...
+        TableTestUtil testUtil = getTestUtil();
+        TableEnvironment tableEnvironment = testUtil.getTableEnv();
+        tableEnvironment.registerCatalog("cat", new CustomCatalog("cat"));
+        tableEnvironment.executeSql(
+                "CREATE TABLE t(i INT, ts TIMESTAMP_LTZ(3), WATERMARK FOR "
+                        + "ts AS ts) WITH ('connector' = 'datagen')");
+        tableEnvironment.executeSql("CREATE VIEW `cat`.`default`.v AS SELECT * FROM t");
+        testUtil.verifyExecPlan(
+                "SELECT sum(i), window_start "
+                        + "FROM TABLE(\n"
```
Comment (snuyanzin) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java:173]: nit: now `TABLE(` is not necessary, could be omitted

**Review by twalthr — APPROVED** [review]
LGTM after addressing the nit comments.

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (dawidwys) [reply flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/catalog/JavaCatalogTableTest.java:162]: The test does not make any sense in non streaming mode. Batch does not deal with time attributes the way streaming does. Nothing fails here.

**Review by snuyanzin — APPROVED** [review]

**Pull request merged**